Civi/Token/ImpliedContextSubscriber.php

```diff
@@ -0,0 +1,105 @@
+<?php
+namespace Civi\Token;
+
+use Civi\Token\Event\TokenRegisterEvent;
+use Civi\Token\Event\TokenValueEvent;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+
+/**
+ * Suppose you have `$context['participantId']`. You can infer a
+ * corresponding `$context['eventId']`. The ImpliedContextSubscriber reads
+ * values like `participantId` and fills-in values like `eventId`.
+ *
+ * @package Civi\Token
+ */
+class ImpliedContextSubscriber implements EventSubscriberInterface {
+
+  public static function getSubscribedEvents() {
+    return [
+      'civi.token.list' => ['onRegisterTokens', 1000],
+      'civi.token.eval' => ['onEvaluateTokens', 1000],
+    ];
+  }
+
+  /**
+   * When listing tokens, ensure that implied data is visible.
+   *
+   * Ex: If `$context['participantId']` is part of the schema, then
+   * `$context['eventId']` will also be part of the schema.
+   *
+   * This fires early during the `civi.token.list` process to ensure that
+   * other listeners see the updated schema.
+   *
+   * @param \Civi\Token\Event\TokenRegisterEvent $e
+   */
+  public function onRegisterTokens(TokenRegisterEvent $e) {
```

Comment:
Non-blocking - but we have been putting return type hints on new functions as well as parameter type hints

Reply:
OK, `void` seems to work here. Squashed/pushed.